signing/signing.go

```diff
@@ -0,0 +1,324 @@
+		log.Println(err1, err2)
+		return
+	}
+
+	// Write response.
+	w.Write([]byte(fmt.Sprintf("Successfully verified and uploaded scorecard results for repo %s on branch %s", []byte(repoName), []byte(repoRef))))
+	w.WriteHeader(http.StatusOK)
+}
+
+func lookupPayload(ctx context.Context, payload []byte) (repoPath, repoRef string, err error) {
+	// Get most recent Rekor entry uuid.
+	rekorClient, err := rekor.NewClient(options.DefaultRekorURL)
+	if err != nil {
+		return "", "", fmt.Errorf("error initializing Rekor client: %v", err)
+	}
+
+	uuids, err := cosign.FindTLogEntriesByPayload(ctx, rekorClient, []byte(payload))
+	if err != nil || len(uuids) == 0 {
+		return "", "", fmt.Errorf("error finding tlog entries corresponding to payload: %v", err)
+	}
+	uuid := uuids[len(uuids)-1] // ignore past entries.
+
+	// Get tlog entry from the UUID.
+	entry, err := cosign.GetTlogEntry(ctx, rekorClient, uuid)
+	if err != nil {
+		return "", "", fmt.Errorf("error fetching tlog entry: %v", err)
+	}
+
+	// Verify tlog entry to make sure it is actually in the log.
+	err = cosign.VerifyTLogEntry(ctx, rekorClient, entry)
+	if err != nil {
+		return "", "", fmt.Errorf("error verifying tlog entry: %v", err)
+	}
+
+	// Extract certificate and get repo reference & path.
+	certs, err := extractCerts(entry)
+	if err != nil || len(certs) == 0 {
+		return "", "", fmt.Errorf("error extracting certificate from entry: %v", err)
+	}
+	if len(certs) > 1 {
+		return "", "", fmt.Errorf("multiple certificates found for the entry: %v", err)
+	}
+
+	cert := certs[0]
+	for _, ext := range cert.Extensions {
+		// OID source: https://github.com/sigstore/fulcio/blob/96ef49cc7662912ba37d46f738757e8d8d5b5355/docs/oid-info.md#L33
+		// TODO: retrieve these by name.
+		if ext.Id.String() == "1.3.6.1.4.1.57264.1.6" {
+			repoRef = string(ext.Value)
+		}
+		if ext.Id.String() == "1.3.6.1.4.1.57264.1.5" {
+			repoPath = string(ext.Value)
+		}
+	}
+
+	if len(repoRef) == 0 || len(repoPath) == 0 {
+		return "", "", fmt.Errorf("error extracting repo ref or path from certificate %v", err)
+	}
+
+	return repoPath, repoRef, nil
+}
+
+func verifyScorecardWorkflow(workflowContent string) error {
+	// Verify workflow contents using actionlint.
+	workflow, lintErrs := actionlint.Parse([]byte(workflowContent))
+	if lintErrs != nil {
+		return fmt.Errorf("actionlint errors parsing workflow: %v", lintErrs)
+	}
+
+	// Extract main job
+	jobs := workflow.Jobs
+	if len(jobs) != 1 {
+		return errors.New("number of jobs isn't 1")
+	}
+	analysisJob := jobs["analysis"]
+	if analysisJob == nil {
+		return errors.New("workflow doens't have analysis job")
+	}
+
+	// Verify that there is no container or services.
+	if analysisJob.Container != nil || len(analysisJob.Services) > 0 {
+		return errors.New("workflow contains container or service")
+	}
+
+	// Verify that the workflow runs on ubuntu-latest and nothing else.
+	if analysisJob.RunsOn != nil {
```

Comment:
re-arrange:
```
if analysisJob.RunsOn == nil {
  return errors.New("no RunsOn found in workflow")
}
```

Note: I think it's allowed for this to be left unspecified and it will default to `ubuntu-latest`. Not 100% sure, please verify

Reply:
Tested it out-- the workflow yml requires the runs-on value to be defined